packages/tangle/booker.go

```diff
@@ -187,71 +167,75 @@ func (b *Booker) BookMessage(messageID MessageID) (err error) {
+	subtractedBranchIDs := pastMarkersBranchIDs.Clone().Subtract(inheritedBranchIDs).Subtract(ledgerstate.NewBranchIDs(ledgerstate.MasterBranchID))
 
-		structureDetails := messageMetadata.StructureDetails()
-		if structureDetails == nil {
-			err = errors.Errorf("failed to retrieve StructureDetails of %s: %w", messageID, cerrors.ErrFatal)
-			return
-		}
-		if structureDetails.PastMarkers.Size() != 1 {
-			err = errors.Errorf("BranchID of %s should have been mapped in the MessageMetadata (multiple PastMarkers): %w", messageID, cerrors.ErrFatal)
-			return
+	if len(addedBranchIDs) != 0 {
+		if aggregatedAddedBranchIDs := b.tangle.LedgerState.AggregateConflictBranchesID(addedBranchIDs); aggregatedAddedBranchIDs != ledgerstate.MasterBranchID {
+			messageMetadata.SetAddedBranchIDs(aggregatedAddedBranchIDs)
 		}
+	}
 
-		branchID = b.MarkersManager.BranchID(structureDetails.PastMarkers.Marker())
-	}) {
-		err = errors.Errorf("failed to load MessageMetadata of %s: %w", messageID, cerrors.ErrFatal)
-		return
+	if len(subtractedBranchIDs) != 0 {
+		if aggregatedSubtractedBranchIDs := b.tangle.LedgerState.AggregateConflictBranchesID(subtractedBranchIDs); aggregatedSubtractedBranchIDs != ledgerstate.MasterBranchID {
+			messageMetadata.SetSubtractedBranchIDs(aggregatedSubtractedBranchIDs)
+		}
 	}
 
-	return
+	return nil
 }
 
-// Shutdown shuts down the Booker and persists its state.
-func (b *Booker) Shutdown() {
-	close(b.shutdown)
-	b.shutdownWG.Wait()
+// determineBookingDetails determines the booking details of an unbooked Message.
+func (b *Booker) determineBookingDetails(message *Message) (parentsStructureDetails []*markers.StructureDetails, parentsPastMarkersBranchIDs, inheritedBranchIDs ledgerstate.BranchIDs, err error) {
+	branchIDsOfPayload, bookingErr := b.bookPayload(message)
+	if bookingErr != nil {
+		return nil, nil, nil, errors.Errorf("failed to book payload of %s: %w", message.ID(), bookingErr)
+	}
 
-	b.MarkersManager.Shutdown()
+	parentsStructureDetails, parentsPastMarkersBranchIDs, strongParentsBranchIDs, bookingDetailsErr := b.collectStrongParentsBookingDetails(message)
+	if bookingDetailsErr != nil {
+		err = errors.Errorf("failed to retrieve booking details of parents of Message with %s: %w", message.ID(), bookingErr)
+		return
+	}
+
+	arithmeticBranchIDs := NewArithmeticBranchIDs(strongParentsBranchIDs.AddAll(branchIDsOfPayload))
+	if weakParentsErr := b.collectWeakParentsBranchIDs(message, arithmeticBranchIDs); weakParentsErr != nil {
+		return nil, nil, nil, errors.Errorf("failed to collect weak parents of %s: %w", message.ID(), weakParentsErr)
+	}
+	if shallowLikeErr := b.collectShallowLikedParentsBranchIDs(message, arithmeticBranchIDs); shallowLikeErr != nil {
+		return nil, nil, nil, errors.Errorf("failed to collect shallow likes of %s: %w", message.ID(), shallowLikeErr)
+	}
+	if shallowDislikeErr := b.collectShallowDislikedParentsBranchIDs(message, arithmeticBranchIDs); shallowDislikeErr != nil {
```

Comment:
These 3 collect methods have a side effect and modify the arithmethic branchIDs which is not clear from the naming and confusing as the `collectStrongParentsBookingDetails` does not have this effect. They should be all consistent in what/how they work. Returning a list and operating with the arithmetic branches in within this function only seems to be cleanest